File: validParentheses_20.py
# // Time Complexity : O(n)
# // Space Complexity : O(n)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No

# // Your code here along with comments explaining your approach

class Solution:
    def isValid(self, s: str) -> bool:
        #stack to keep track of opening brackets
        stack = []
        
        #Hash map for keeping track of bracket mappings
        hmap = {')':'(', 
                '}':'{', 
                ']':'['}
        
        # If the character is an closing bracket, Pop the topmost element from the stack, if it is non empty
        # else, assign a dummy value of '#' to the top_element variable
        for char in s:
            if char in hmap:
                top = stack.pop() if stack else '#'
                # if mapping don't match, return False
                if hmap[char] != top:
                    return False
            else:
                #an opening bracket: push it onto the stack
                stack.append(char)
                
        return True if not stack else False

File: Valid_Parentheses_20.java
//Time Complexity : O(n) // n : s.length()
//Space Complexity : O(n) // n : s.length()
//Did this code successfully run on Leetcode : Yes
//Any problem you faced while coding this : None

package com.s30.satish;
import java.util.Stack;

class Valid_Parentheses_20 {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        for(int i = 0; i < s.length(); i++)
        {
            char ch = s.charAt(i);
            if(ch == '(' || ch == '[' || ch == '{')
                st.push(ch);
            else if(ch == ')' || ch == ']' || ch == '}')
            {
                if(!st.isEmpty())
                {
                    char poppedChar = st.pop();
                    if((poppedChar == '(' && ch != ')') || 
                       (poppedChar == '[' && ch != ']') ||
                       (poppedChar == '{' && ch != '}'))
                        return false;
                }
                else if(st.isEmpty())
                	return false;
            }
        }
        if(!st.isEmpty())
            return false;
        else
            return true;
    }
    public static void main(String[] args)
    {
    	Valid_Parentheses_20 para = new Valid_Parentheses_20();
    	System.out.println(para.isValid("(){}[]"));
    	System.out.println(para.isValid("({[]})"));
    	System.out.println(para.isValid("(){}}{"));
    	System.out.println(para.isValid("{"));
    }
}